### PR TITLE
Adding upgrade test for TestCustomGetQuery() in graphql/e2e/custom_logic

### PR DESCRIPTION
**Pull request by SiddeshUbale**
**Adding upgrade test for TestCustomGetQuery() in graphql/e2e/custom_logic.**

**Problem:-**
1] Added test only for TestCustomGetQuery() function from graphql/e2e/custom_logic.
2] Comments and commented code lines are not yet removed.
3] Some logs which are useful for development are not yet removed.

### CODE REVIEW

File: contrib/Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # This gets built as part of release.sh. Must be run from /tmp/build, with the linux binaries
 # already built and placed there.
 
-FROM ubuntu:20.04
```
Comment (mangalaman93) [inline contrib/Dockerfile:7]: undo these changes

File: dgraphtest/dgraph.go [hunk 2]
```diff
@@ -290,6 +290,11 @@ func publicPort(dcli *docker.Client, dc dnode, privatePort string) (string, erro
 	return "", fmt.Errorf("no mapping found for private port [%v] for container [%v]", privatePort, dc.cname())
 }
 
+// Wrapper function to export publicport() in other files
```
Comment (mangalaman93) [inline dgraphtest/dgraph.go:293]: we shouldn't add this function

File: dgraphtest/graphql.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+func RequireNoError(err error) {
+	if err != nil {
+		debug.PrintStack()
+		panic(fmt.Sprintf("required no GraphQL errors, but received: %s\n", err.Error()))
+	}
+}
+
+func Equal(expected interface{}, actual interface{}) bool {
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:72]: Does require.Equal not work?

File: dgraphtest/graphql.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	// If the types are different, they are not equal
+	if value1.Type() != value2.Type() {
+		return false
+	}
+
+	// Compare the values of the interfaces
+	return reflect.DeepEqual(expected, actual)
+}
+
+func (hc *HTTPClient) SafelyUpdateGQLSchema(schema string, alphaPort string) *GqlSchema {
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:86]: We already have UpdateGQLSchema function that is already available in dgraphtest package?

File: dgraphtest/graphql.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+// **********************************************************************************************************
+
+func (hc *HTTPClient) probeGraphQL(alphaPort string) (*ProbeGraphQLResp, error) {
+
+	request, err := http.NewRequest("GET", "http://localhost:"+alphaPort+"/probe/graphql", nil)
+	if err != nil {
+		return nil, err
+	}
+	client := &http.Client{}
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:111]: we should not create a new HTTP client, we should use the existing global variable.

File: dgraphtest/graphql.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+		resp, err := hc.probeGraphQL(alphaPort)
+		if err == nil && resp.Healthy {
+			return resp
+		}
+		time.Sleep(time.Second)
+	}
+	return nil
+}
+
+func (hc *HTTPClient) RetryProbeGraphQL(alphaPort string) *ProbeGraphQLResp {
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:144]: we probably do not need this function. This could be replaced `retryProbeGraphQL` function

File: dgraphtest/graphql.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	//RequireNoGQLErrors(updateResp)
+
+	// sanity: update response should reflect the new schema
+	var updateResult struct {
+		UpdateGQLSchema struct {
+			GqlSchema *GqlSchema
+		}
+	}
+	if err := json.Unmarshal(updateResp, &updateResult); err != nil {
+		debug.PrintStack()
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:170]: when we panic, we do not need to call debug.PrintStack()

File: dgraphtest/graphql.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		panic(fmt.Sprintf("failed to unmarshal updateGQLSchema response: %s", err.Error()))
+	}
+	if updateResult.UpdateGQLSchema.GqlSchema == nil {
+		panic(fmt.Sprintf("Expected value not to be nil."))
+	}
+	_, err := dql.ParseUid(updateResult.UpdateGQLSchema.GqlSchema.Id)
+	RequireNoError(err)
+	// require.Equalf(t, updateResult.UpdateGQLSchema.GqlSchema.Schema, schema,
+	// 	"updateGQLSchema response doesn't reflect the updated schema")
+	if !Equal(updateResult.UpdateGQLSchema.GqlSchema.Schema, schema) {
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:180]: why are we not using require.NotEqual?

File: dgraphtest/graphql.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+			return resp
+		}
+
+		// otherwise, retry schema update
+		panic(fmt.Sprintf("Got error while updateGQLSchema: %s. Retrying...\n", err.Error()))
+		time.Sleep(time.Second)
+	}
+}
+
+func (hc *HTTPClient) updateGQLSchema(alphaPort string, schema string) ([]byte, error) {
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:204]: this function is already present.

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

File: contrib/Dockerfile (hunk 1 above)
Comment (SiddeshUbale) [reply contrib/Dockerfile:7]: Resolved

File: dgraphtest/dgraph.go (hunk 2 above)
Comment (SiddeshUbale) [reply dgraphtest/dgraph.go:328]: Resolved

File: dgraphtest/graphql.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+func RequireNoGQLErrors(resp *GraphQLResponse) {
+	if resp != nil {
+		panic(fmt.Sprintf("Expected value not to be nil."))
+	}
+	RequireNoError(resp.Errors)
+}
+
+func RequireNoError(err error) {
+	if err != nil {
+		debug.PrintStack()
```
[the remark this answers is not in the document]
Comment (SiddeshUbale) [reply dgraphtest/graphql.go:67]: Resolved

File: dgraphtest/graphql.go (hunk 4 above)
Comment (SiddeshUbale) [reply dgraphtest/graphql.go:86]: Resolved

File: dgraphtest/graphql.go (hunk 6 above)
Comment (SiddeshUbale) [reply dgraphtest/graphql.go:144]: Resolved

File: dgraphtest/graphql.go (hunk 7 above)
Comment (SiddeshUbale) [reply dgraphtest/graphql.go:170]: Resolved

File: dgraphtest/graphql.go (hunk 9 above)
Comment (SiddeshUbale) [reply dgraphtest/graphql.go:204]: Resolved

File: graphql/e2e/custom_logic/custom_logic_test.go [hunk 11]
```diff
@@ -78,37 +80,37 @@ const (
...
-		Query: query,
-	}
-
-	result := params.ExecuteAsPost(t, common.GraphqlURL)
-	common.RequireNoGQLErrors(t, result)
-
-	expected := `{"myFavoriteMovies":[{"id":"0x3","name":"Star Wars","director":[{"id":"0x4","name":"George Lucas"}]},{"id":"0x5","name":"Star Trek","director":[{"id":"0x6","name":"J.J. Abrams"}]}]}`
-	require.JSONEq(t, expected, string(result.Data))
-}
+// func TestCustomGetQuery(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (SiddeshUbale) [reply graphql/e2e/custom_logic/custom_logic_test.go]: Resolved

File: dgraphtest/dgraph.go (hunk 2 above)
Comment (mangalaman93) [reply dgraphtest/dgraph.go:328]: how is it resolved? The function is still exported

File: dgraphtest/graphql.go [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/dgraph-io/dgraph/dql"
+	"github.com/dgraph-io/dgraph/x"
+	"github.com/pkg/errors"
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:28]: fix imports order

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

File: dgraphtest/config.go [hunk 13]
```diff
@@ -69,10 +69,15 @@ var AllUpgradeCombos = []UpgradeCombo{
...
 
+type genContainerConfig struct {
+	buildContext string
+}
+
 type ClusterConfig struct {
 	prefix         string
 	numAlphas      int
 	numZeros       int
+	genContainers  int
```
Comment (mangalaman93) [inline dgraphtest/config.go:80]: nit: genCs is also a good name

File: dgraphtest/config.go [hunk 14]
```diff
@@ -69,10 +69,15 @@ var AllUpgradeCombos = []UpgradeCombo{
 	{"c5862ae2a", "v23.0.0", BackupRestore},
 }
 
+type genContainerConfig struct {
```
Comment (mangalaman93) [inline dgraphtest/config.go:72]: nit: genCConfig

File: dgraphtest/dgraph.go [hunk 15]
```diff
@@ -90,6 +93,86 @@ type dnode interface {
...
+	return zcmd
+}
+
+func (gen *genericContainer) workingDir() string {
+	return genericWorkingDir
+}
+
+func (gen *genericContainer) mounts(c *LocalCluster) ([]mount.Mount, error) {
+	var mounts []mount.Mount
+	binMount, err := mountBinary(c)
```
Comment (mangalaman93) [inline dgraphtest/dgraph.go:144]: why are mounting dgraph binary?

File: dgraphtest/dgraph.go [hunk 16]
```diff
@@ -90,6 +93,86 @@ type dnode interface {
...
+	if err != nil {
+		return nil, err
+	}
+	mounts = append(mounts, binMount)
+	return mounts, nil
+}
+
+func (gen *genericContainer) healthURL(c *LocalCluster) (string, error) {
+	publicPort, err := publicPort(c.dcli, gen, genericPort)
```
Comment (mangalaman93) [inline dgraphtest/dgraph.go:153]: health URL won't exist for genC

File: dgraphtest/dgraph.go [hunk 17]
```diff
@@ -90,6 +93,86 @@ type dnode interface {
...
+	if err != nil {
+		return "", err
+	}
+	return "http://localhost:" + publicPort + "/health", nil
+}
+
+func (gen *genericContainer) assignURL(c *LocalCluster) (string, error) {
```
Comment (mangalaman93) [inline dgraphtest/dgraph.go:160]: same

File: dgraphtest/graphql.go [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+	SchemaUpdateCounter uint64
+}
+
+type GraphQLResponse struct {
+	Data       json.RawMessage        `json:"data,omitempty"`
+	Errors     x.GqlErrorList         `json:"errors,omitempty"`
+	Extensions map[string]interface{} `json:"extensions,omitempty"`
+}
+
+var safelyUpdateGQLSchemaErr = "New Counter: %v, Old Counter: %v.\n" +
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:49]: do we need this?

File: dgraphtest/graphql.go [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+	"Schema update counter didn't increment, " +
+	"indicating that the GraphQL layer didn't get the updated schema even after 10" +
+	" retries. The most probable cause is the new GraphQL schema is same as the old" +
+	" GraphQL schema."
+
+func RequireNoError(err error) {
```
Comment (mangalaman93) [inline dgraphtest/graphql.go:55]: we already have x.Panic and other functions, we don't need this

File: dgraphtest/graphql.go (hunk 4 above)
Comment (mangalaman93) [reply dgraphtest/graphql.go:75]: I still see this function here

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
reviewed most of the PR. Can you cleanup the rest of the PR and address comments, I will review it again then.
